In [1]:
import pandas as pd

In [2]:
%run Immo_Assistant.app.ipynb

<>:2: SyntaxWarning: invalid escape sequence '\L'
C:\Users\Weverson Barbieri\AppData\Local\Temp\ipykernel_11920\507487594.py:2: SyntaxWarning: invalid escape sequence '\L'
  df = pd.read_excel('C:\Language_Projects\Language_Projects\Python\Flagship_1\Immo_Assistant.app\data\year_make_model_df.xlsx')


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3463 entries, 0 to 3462
Data columns (total 5 columns):
 #   Column          Non-Null Count  Dtype 
---  ------          --------------  ----- 
 0   Year            3463 non-null   int64 
 1   Make            3463 non-null   object
 2   Model           3463 non-null   string
 3   Security        3156 non-null   object
 4   ParameterReset  938 non-null    object
dtypes: int64(1), object(3), string(1)
memory usage: 135.4+ KB
PATS Type A (Stand Alone PATS Module)
PATS Type A (Stand Alone PATS Module)
PATS Type A (Stand Alone PATS Module)
PATS Type A (Stand Alone PATS Module)
PATS Type A (Stand Alone PATS Module)
Built February 1st or earlier: PATS Type A (Stand Alone PATS Module)
Built February 2nd or later: PATS Type E (Powertrain Control Module)
PATS Type B (Stand Alone PATS Module)
PATS Type A (Stand Alone PATS Module)
PATS Type B (Stand Alone PATS Module)
PATS Type B (Stand Alone PATS Module)
PATS Type B (Stand Alone PATS Module)
PATS 

C:\Users\Weverson Barbieri\AppData\Local\Temp\ipykernel_11920\3384685316.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_ford[['PATS Type', 'PATS Module Location']] = df_ford['Security'].str.split('(', n=1, expand=True)
C:\Users\Weverson Barbieri\AppData\Local\Temp\ipykernel_11920\3384685316.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_ford[['PATS Type', 'PATS Module Location']] = df_ford['Security'].str.split('(', n=1, expand=True)
C:\Users\Weverson Barbieri\AppData\Local\Temp\ipykernel_11

Year: 2005
Make: Cadillac
Model: CTS
Security: PK3+
Hardware Part Number: U44G-G45
Is this hardware clonable?: No
['Pontiac', 'Mazda', 'Acura', 'Honda', 'BMW', 'Lexus', 'Toyota', 'Chrysler', 'Dodge', 'Infiniti', 'Jaguar', 'Lincoln', 'Nissan', 'Plymouth', 'Jeep', 'Audi', 'Mitsubishi', 'Volkswagen', 'Land Rover', 'Mini', 'Rolls-Royce', 'Scion', 'Subaru', 'Fiat']
['Pontiac', 'Mazda', 'Acura', 'Honda', 'BMW', 'Lexus', 'Toyota', 'Chrysler', 'Dodge', 'Infiniti', 'Jaguar', 'Lincoln', 'Nissan', 'Plymouth', 'Jeep', 'Audi', 'Mitsubishi', 'Volkswagen', 'Land Rover', 'Mini', 'Rolls-Royce', 'Scion', 'Subaru', 'Fiat']


<>:2: SyntaxWarning: invalid escape sequence '\L'
C:\Users\Weverson Barbieri\AppData\Local\Temp\ipykernel_11920\1252695280.py:2: SyntaxWarning: invalid escape sequence '\L'
  df_ford_last.to_csv('C:\Language_Projects\Language_Projects\Python\Flagship_1\Immo_Assistant.app\dataframe_each_make_xlsx_file\df_ford.csv', index=False)
<>:2: SyntaxWarning: invalid escape sequence '\L'
C:\Users\Weverson Barbieri\AppData\Local\Temp\ipykernel_11920\1035310976.py:2: SyntaxWarning: invalid escape sequence '\L'
  df_mercury_last.to_csv('C:\Language_Projects\Language_Projects\Python\Flagship_1\Immo_Assistant.app\dataframe_each_make_xlsx_file\df_mercury.csv', index=False)
<>:2: SyntaxWarning: invalid escape sequence '\L'
C:\Users\Weverson Barbieri\AppData\Local\Temp\ipykernel_11920\526802729.py:2: SyntaxWarning: invalid escape sequence '\L'
  df_chevrolet_final_verson.to_csv('C:\Language_Projects\Language_Projects\Python\Flagship_1\Immo_Assistant.app\dataframe_csv_files\df_chevy_filtered.csv', index=Fa

In [3]:
# Calling the function
df_bmw = create_make_df('BMW')

In [4]:
# Convert all models as integer type to string type 
df_bmw = df_bmw.astype({'Model': 'string'})

In [5]:
# Check if there are null values.
df_bmw.info()

<class 'pandas.core.frame.DataFrame'>
Index: 266 entries, 119 to 3213
Data columns (total 5 columns):
 #   Column          Non-Null Count  Dtype 
---  ------          --------------  ----- 
 0   Year            266 non-null    int64 
 1   Make            266 non-null    object
 2   Model           266 non-null    string
 3   Security        266 non-null    object
 4   ParameterReset  245 non-null    object
dtypes: int64(1), object(3), string(1)
memory usage: 12.5+ KB


In [6]:
# Create a list to store the elemets from column 'Model'
bmw_elements_model = []
# Create a list to store unique elements
bmw_unique_elements_model = []

# Loop to get all elements from the said column 'Model'
for model in df_bmw['Model']:
    bmw_elements_model.append(model)
    if model not in bmw_unique_elements_model:
        bmw_unique_elements_model.append(model)

In [7]:
# Print the list w/ unique values
print(bmw_unique_elements_model)

['318', '325', '525', '530', '535', '540', '840', '850', '740/745*', '750/760*', 'M3', '328', 'Z3', '528', '323', 'M Coupe/Roadster', 'M5', 'X5', '330', 'Z8', '745', '760', 'Z4', '645ci', 'X3', '545', '750', '335', '550', '650', 'M6', 'Z4M coupe & M Roadster', '1 Series (128i/135i)', '3 Series Cabriolet\nretractable hardtop', 'X6', '128', '135', '640', '740', 'Alpina B7', 'X1']


1. Checking the list w/ all BMW models, these are the ones which need to be changed. As there are 2 models in the same row, they will be separated: 
- 740/745*, 
- 750/760*, 
- 1 Series (128i/135i)

2. The above models will be separated and they should appear as below:
- 740
- 745
- 750
- 760
- 128i
- 135i

The models below do not need to be separated as they are only the type of body:
- M Coupe/Roadster

In [8]:
# Create a list w/ the BMW models containing the caractere '/' 
bmw_model_change_list = []
# Create a list w/ the indexes 
bmw_model_change_index_list = []

# Loop to get all models containing the caractere '/', except model 'M Coupe/Roadster'
for index, row in df_bmw.iterrows():
    if '/' in row['Model'] and row['Model'] != 'M Coupe/Roadster':
        bmw_model_change_list.append(row)
        bmw_model_change_index_list.append(index)

In [9]:
# Convert the list of models w/ the caractere '/' to dataframe 
df_bmw_to_change = pd.DataFrame(bmw_model_change_list)

In [10]:
df_bmw_to_change

Year Make                 Model Security ParameterReset
127   1995  BMW              740/745*      Std         EWS II
128   1995  BMW              750/760*      Std         EWS II
170   1996  BMW              740/745*      Std         EWS II
171   1996  BMW              750/760*      Std         EWS II
235   1997  BMW              740/745*      Std        EWS III
236   1997  BMW              750/760*      Std        EWS III
299   1998  BMW              740/745*      Std        EWS III
300   1998  BMW              750/760*      Std        EWS III
402   1999  BMW              740/745*      Std        EWS III
403   1999  BMW              750/760*      Std        EWS III
517   2000  BMW              740/745*      Std        EWS III
518   2000  BMW              750/760*      Std        EWS III
668   2001  BMW              740/745*      Std        EWS III
669   2001  BMW              750/760*      Std        EWS III
825   2002  BMW              740/745*      Std        EWS III
826   2002  BMW              750/760*      Std        EWS III
2005  2008  BMW  1 Series (128i/135i)      Std        EWS III
2247  2009  BMW  1 Series (128i/135i)      Std        EWS III
2503  2010  BMW  1 Series (128i/135i)      Std        EWS III
2742  2011  BMW  1 Series (128i/135i)      Std        EWS III
2971  2012  BMW  1 Series (128i/135i)      Std        EWS III

In [11]:
# Separate/split the models, using the caractere '/' as delimiter
df_bmw_to_change = df_bmw_to_change.assign(Model=df_bmw_to_change['Model'].str.split('/')).explode('Model')

In [12]:
# Create a list to store all rows containing the string and caracteres '1 Series (128i', '135i)', '*'
bmw_remove_char_list = []
# Create a list with all indexes
bmw_remove_char_index_list = []

# Loop into column 'Model', get those rows with the string '1 Series (128i'
for index, row in df_bmw_to_change.iterrows():
    if '1 Series (128i' in row['Model']:
        # Save in the above list
        bmw_remove_char_list.append(row)
        bmw_remove_char_index_list.append(index)

# Change the list to dataframe
df_bmw_remove_char = pd.DataFrame(bmw_remove_char_list)

In [13]:
# Remove the string '1 Series (' and keep only the string '128i'.
# The code '...['Model'].replace({"1 Series (": ""}...' generated a error. It was needed to add \\ (... 1 Series \\("...) to fix the error
df_bmw_remove_char['Model'] = df_bmw_remove_char['Model'].replace({"1 Series \\(": ""}, regex=True)

In [14]:
# Confirm if the changes were properly done
df_bmw_remove_char.head(2)

Year Make Model Security ParameterReset
2005  2008  BMW  128i      Std        EWS III
2247  2009  BMW  128i      Std        EWS III

In [15]:
# Concatenate the main df_bmw to the df_bmw_remove_char
df_bmw_changes_done = pd.concat([df_bmw_to_change, df_bmw_remove_char])

In [16]:
# Concatenate the main df_bmw with the df after changes done
df_bmw_concatenated = pd.concat([df_bmw_changes_done, df_bmw])

In [17]:
# Reset index
df_bmw_reset_index = df_bmw_concatenated.reset_index(drop=True)

In [18]:
# Create a list with all values from column model
bmw_model_column_strings = []

# Loop to create the above list from column 'Model'
for value in df_bmw_reset_index['Model']:
    bmw_model_column_strings.append(value)

In [19]:
# Create a list w/ unique values from column Model
bmw_model_unique_string = set(list(bmw_model_column_strings))

In [20]:
# Print the list to confirm the changes that need to be done
print(bmw_model_unique_string)

{'530', '128i', '528', '650', 'X5', '645ci', '740', 'Alpina B7', '525', '750', 'Z3', '3 Series Cabriolet\nretractable hardtop', '745*', '850', 'Z8', '318', 'M5', '745', 'M Coupe/Roadster', 'X3', '760*', 'Z4', '750/760*', '760', '330', '1 Series (128i/135i)', '840', '325', '135i)', '323', '540', '335', '545', '135', '1 Series (128i', '128', 'X1', '640', '328', '535', 'M6', '740/745*', 'M3', 'Z4M coupe & M Roadster', 'X6', '550'}


* CHANGES NEEDED:
- Remove the caracteres '*' and ')', '\n'
- Remove the models '1 Series (128i/135i)', '1 Series (128i', '740/745*', '750/760*'

In [21]:
# Create a list from the indexes w/ rows containing the string '1 Series'
bmw_indexes_to_remove_first_list = []

# Loop to get the indexes
for index, row in df_bmw_reset_index.iterrows():
    if '1 Series' in row['Model']:
        bmw_indexes_to_remove_first_list.append(index)

In [22]:
# Print the list
bmw_indexes_to_remove_first_list

[32, 34, 36, 38, 40, 207, 225, 243, 261, 282]

In [23]:
# Create a list from the indexes w/ rows containing the string '1 Series'
bmw_indexes_to_remove_second_list = []

# Loop to get the indexes
for index, row in df_bmw_reset_index.iterrows():
    if '/' in row['Model'] != 'M Coupe/Roadster':
        bmw_indexes_to_remove_second_list.append(index)

In [24]:
# Print the list
print(bmw_indexes_to_remove_second_list)

[55, 56, 63, 64, 73, 74, 82, 83, 92, 93, 101, 102, 114, 115, 127, 128, 207, 225, 243, 261, 282]


In [25]:
# Use this list comprehesion to compare the indexes from first_list and the ones from the second_list. Adding indexes the second list does not have
# from the first list and remove the duplicate. The goal is to keep the second list as the main   
[bmw_indexes_to_remove_second_list.append(index) for index in bmw_indexes_to_remove_first_list if index not in bmw_indexes_to_remove_second_list]

[None, None, None, None, None]

In [26]:
# Print the second list (main)
print(bmw_indexes_to_remove_second_list)

[55, 56, 63, 64, 73, 74, 82, 83, 92, 93, 101, 102, 114, 115, 127, 128, 207, 225, 243, 261, 282, 32, 34, 36, 38, 40]


In [27]:
# Drop all indexes stored in the list
df_bmw_column_model_first_change = df_bmw_reset_index.drop(bmw_indexes_to_remove_second_list)

In [28]:
# Remove the caracteres '*' and ')', '\n'
df_bmw_column_model_first_change['Model'] = df_bmw_column_model_first_change['Model'].str.replace(")", "", regex=False).str.replace("*", "", regex=False).str.replace("\n", " ", regex=False)

In [29]:
# Function to create a list w/ unique values from the columns 
def strings_from_column(df_make, column):
    # List to store the strings from the columns
    strings_from_column_list = []
    # Loop to get the strings from each row based on the column given
    for string in df_make[column]:
        # Store the strings in the list
        strings_from_column_list.append(string)
    # Return a list w/ unique values/strings from the column given
    return set(strings_from_column_list)

In [30]:
# Call the function to show a list with unique values containing all strings from column model
bmw_model_unique_string_list_updated = strings_from_column(df_bmw_column_model_first_change, 'Model')

In [31]:
# Print the list to be analysed and see if the models are correct after the updates done
print(bmw_model_unique_string_list_updated)

{'530', '128i', '528', '650', 'X5', '645ci', '740', 'Alpina B7', '525', '750', 'Z3', '850', 'Z8', '318', 'M5', '745', 'M Coupe/Roadster', 'X3', 'Z4', '3 Series Cabriolet retractable hardtop', '760', '330', '840', '325', '323', '540', '335', '545', '135', '128', 'X1', '640', '328', '535', 'M6', 'M3', 'Z4M coupe & M Roadster', 'X6', '135i', '550'}


In [32]:
# Check if there are null values in the column 'ParameterReset'
df_bmw_column_model_first_change.info()

<class 'pandas.core.frame.DataFrame'>
Index: 287 entries, 0 to 312
Data columns (total 5 columns):
 #   Column          Non-Null Count  Dtype 
---  ------          --------------  ----- 
 0   Year            287 non-null    int64 
 1   Make            287 non-null    object
 2   Model           287 non-null    object
 3   Security        287 non-null    object
 4   ParameterReset  266 non-null    object
dtypes: int64(1), object(4)
memory usage: 13.5+ KB


In [33]:
# Create a list w/ strings/values from column security
bmw_security_list = strings_from_column(df_bmw_column_model_first_change, 'Security')

In [34]:
# Print the list
bmw_security_list

{'Std'}

No need to change the values from column 'Security', as the entire column has the string 'Std' indicating the security system/theft system is an standard equipment for all BMW models.

In [35]:
# Create a list w/ strings/values from column ParameterReset
bmw_pr_list = strings_from_column(df_bmw_column_model_first_change, 'ParameterReset')

In [36]:
# Print the list
bmw_pr_list

{'EWS II', 'EWS III', nan}

CHANGES NEEDED:
- Fill out null values.
- Rename ParameterReset column to 'BMW Security System Additional Information'

In [37]:
# Fill out null values
df_bmw_column_model_first_change['ParameterReset'].fillna('Information not avaiable', inplace=True)

C:\Users\Weverson Barbieri\AppData\Local\Temp\ipykernel_11920\4162266557.py:2: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df_bmw_column_model_first_change['ParameterReset'].fillna('Information not avaiable', inplace=True)


In [38]:
# Rename ParameterReset column to 'BMW Security System Additional Information'
df_bmw_updated = df_bmw_column_model_first_change.rename(columns={'ParameterReset':'BMW Security System Additional Information'})

In [39]:
# Check the df after updating
df_bmw_updated.info()

<class 'pandas.core.frame.DataFrame'>
Index: 287 entries, 0 to 312
Data columns (total 5 columns):
 #   Column                                      Non-Null Count  Dtype 
---  ------                                      --------------  ----- 
 0   Year                                        287 non-null    int64 
 1   Make                                        287 non-null    object
 2   Model                                       287 non-null    object
 3   Security                                    287 non-null    object
 4   BMW Security System Additional Information  287 non-null    object
dtypes: int64(1), object(4)
memory usage: 13.5+ KB


In [40]:
# Check the head of the df to confirm if the index is reset
df_bmw_updated.head(5)

Year Make Model Security BMW Security System Additional Information
0  1995  BMW   740      Std                                     EWS II
1  1995  BMW   745      Std                                     EWS II
2  1995  BMW   750      Std                                     EWS II
3  1995  BMW   760      Std                                     EWS II
4  1996  BMW   740      Std                                     EWS II

In [41]:
# Download the BMW df to .pkl file
df_bmw_updated.to_excel('df_bmw_updated.xlsx')